src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs

```diff
@@ -710,6 +708,50 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
             }
 
             return UpdateExpression(builder, node.Update(left, right, node.IsRef, node.Type));
+
+            BoundExpression spillReceiver(BoundFieldAccess field, ref BoundSpillSequenceBuilder leftBuilder, bool isFieldLeft)
+            {
+                BoundExpression receiver;
+                if (field.FieldSymbol.ContainingType.IsReferenceType)
+                {
+                    // a reference type can always live across await so Spill using leftBuilder
+                    receiver = Spill(leftBuilder, VisitExpression(ref leftBuilder, field.ReceiverOpt));
+                }
+                else if (field.ReceiverOpt is BoundArrayAccess arrayAccess)
+                {
+                    // an arrayAccess returns a ref so can only be called after the await, but spill expression and indices
+                    var expression = VisitExpression(ref leftBuilder, arrayAccess.Expression);
+                    expression = Spill(leftBuilder, expression, RefKind.None);
+                    var indices = this.VisitExpressionList(ref leftBuilder, arrayAccess.Indices, forceSpill: true);
+                    receiver = arrayAccess.Update(expression, indices, arrayAccess.Type);
+                    // dummy array access to trigger IndexOutRangeException or NRE
+                    // we only need this if the array access is a receiver since
+                    // a[0] = b triggers a NRE/IORE on assignment
+                    // but a[0].b = c triggers an NRE/IORE before evaluating c
+                    Spill(leftBuilder, receiver, sideEffectsOnly: true);
```

Comment:
>Spill(leftBuilder, receiver, sideEffectsOnly: true); [](start = 20, length = 52)

Are we going to evaluate this multiple times when this is a receiver of a not "isFieldLeft" field?